### PR TITLE
add ticket authentication support [closes #66]

### LINKED ISSUE #66: Example: Ticket authentication
Hello,
for beginning, thank you for this package.

I've just implemented wampy to our stack, and I have some code that I think would be useful for readme or examples.
In our stack, Crossbar router runs in separate docker container, and is configured to use ticket authentication (which is not natively supported in wampy).

First, I created a new `TicketAuthenticate` class - message that will be sent after `ticket` challenge.
It receives `ticket` in constructor.
```py
class TicketAuthenticate(object):
    WAMP_CODE = 5
    name = "authenticate"

    def __init__(self, ticket, kwargs_dict=None):
        super(TicketAuthenticate, self).__init__()

        self.ticket = ticket
        self.kwargs_dict = kwargs_dict or {}

    @property
    def message(self):
        return [
            self.WAMP_CODE, self.ticket, self.kwargs_dict
        ]
```

Then, I extended `MessageHandler` to respond with this message in `handle_challenge`. It receives ticket in constructor and passes it to `TicketAuthenticate` once requested.

```py
from wampy.message_handler import MessageHandler, logger

class TicketMessageHandler(MessageHandler):

    ticket = None

    def __init__(self, ticket) -> None:
        super().__init__()
        self.ticket = ticket

    def handle_challenge(self, message_obj):
        logger.debug("client has been Challenged")
        message = TicketAuthenticate(self.ticket)
        self.session.send_message(message)
```

And finally usage example
```py
        roles = {
            'roles': {
                'subscriber': {},
                'publisher': {},
                'callee': {
                    'shared_registration': True,
                },
                'caller': {},
            },
            'authmethods': ['ticket'],
            'authid': settings.WAMP_BACKEND_PRINCIPAL
        }
        self.client = Client(
            url=f"ws://{settings.WAMP_BACKEND_HOST}:{settings.WAMP_BACKEND_PORT}",
            realm=settings.WAMP_BACKEND_REALM,
            roles=roles,
            message_handler=TicketMessageHandler(settings.WAMP_BACKEND_TICKET)
        )
        self.client.start()
```

I'm not a Python developer, but I feel that this code is not the best approach - that's why I'm creating this issue - it's not suitable for a pull request. 
It could save some time to anyone though, maybe it would be an inspiration for adding a support for ticket auth to wampy.

Please let me know whether you want this issue to be closed without discussion (just kept for reference), or you'd like to discuss this topic.

Thanks!

### PR DESCRIPTION
**Pull request by wodCZ**
Hi,
here is my first attempt to implement ticket authentication to wampy.

Unfortunately some tests have been failing for me right when I cloned this repository, so I'm not 100% sure I didn't break something else.

Please let me know if there are some issues.

Thanks!

### CODE REVIEW

**Comment by wodCZ** [comment]
Uh, all red. Seems to be caused by different versions of installed libraries.

**Comment by noisyboiler** [comment]
the first of many, i hope!
many thanks @wodCZ

**Comment by wodCZ** [comment]
You're welcome @noisyboiler - we use this library in production, so it's possible more PRs are comming :)

Freezing the Twisted dev dependency at 17.9 fixed most tests - these were failing because of incompatibility with crossbar/autobahn.

Failing Python 3.3 tests - [Setuptools dropped python 3.3 support in v40](https://setuptools.readthedocs.io/en/latest/history.html#v40-0-0).

I'm not sure what to do about that.

**Comment by wodCZ** [comment]
Skipping the setuptools upgrade step for python 3.3 using
```
  - if [[ $TRAVIS_PYTHON_VERSION != '3.3' ]]; then pip install --upgrade setuptools; fi
``` 
could fix the 3.3 build.

What do you think @noisyboiler ?

**Comment by noisyboiler** [comment]
Hey. Sorry for the slow replies but I'll do my best to address this tomorrow or the coming days.

File: setup.py [hunk 1]
```diff
@@ -44,6 +44,7 @@
         'dev': [
             "crossbar==0.15.0",
             "autobahn==0.17.2",
+            "Twisted==17.9.0",
```
Comment (noisyboiler) [inline setup.py]: 👍

**Comment by noisyboiler** [comment]
i'm back!

File: wampy/message_handler.py [hunk 2]
```diff
@@ -75,10 +75,15 @@ def handle_challenge(self, message_obj):
             return
 
         secret = os.environ['WAMPYSECRET']
-        challenge_data = message_obj.challenge
-        signature = compute_wcs(secret, str(challenge_data))
+        if message_obj.auth_method == 'ticket':
+            logger.debug("proceeding with ticket authentication method")
+            message = Authenticate(secret)
+        else:
```
Comment (noisyboiler) [inline wampy/message_handler.py]: I'm happy enough to merge this as it is, but now you've made me aware of the different auth methods `wampy` should support, I'd eventually prefer to see an `elseif` here conditioning against `wampcra`, then the hanging `else` raising a `RuntimeError` or something, as at some point someone else will play about with another auth method. I'll give you some time to think.

Also, going forward I'd like to use `Enum` to martial these strings into specific types. So something like...
```
class AuthMethod(Enum):
    ticket = 'ticket'
    cra = 'wampcra'
```
would look great in `wampy.constants.py`!

**Comment by noisyboiler** [comment]
@wodCZ 
much appreciated!
to fix your tests just merge in master.

File: wampy/message_handler.py (hunk 2 above)
Comment (wodCZ) [reply wampy/message_handler.py]: My point here was to add explicit support for the `ticket` auth method while maintaining backwards compatibility as much as possible.

I agree with the Enum and `RuntimeError` approach, but I'm not sure when I could get into it. Possibly within a month, but you know how it goes 😀

**Comment by wodCZ** [comment]
Master merged, tests are fine now, thanks 🎉 

As I wrote in https://github.com/noisyboiler/wampy/pull/68#discussion_r228887959 I'm not sure when I could get into Enum refactoring, so I'd prefer to merge this and leave the refactoring for another Pull Request, possibly with support for other authentication methods.

**Comment by noisyboiler** [comment]
happy to merge. I'll add your name to the contributors file.
thank you @wodCZ

**Pull request merged**